elements/lisk-transactions/src/15_proof_of_misbehavior_transaction.ts

```diff
@@ -0,0 +1,442 @@
+					type: 'integer',
+					minimum: 0,
+				},
+				payloadLength: {
+					type: 'integer',
+					minimum: 0,
+				},
+				previousBlockId: {
+					type: ['null', 'string'],
+					format: 'id',
+					minLength: 1,
+					maxLength: 20,
+				},
+				generatorPublicKey: {
+					type: 'string',
+					format: 'publicKey',
+				},
+				maxHeightPrevoted: {
+					type: 'integer',
+					minimum: 0,
+				},
+				maxHeightPreviouslyForged: {
+					type: 'integer',
+					minimum: 0,
+				},
+				height: {
+					type: 'integer',
+					minimum: 1,
+				},
+				blockSignature: {
+					type: 'string',
+					format: 'signature',
+				},
+			},
+		},
+	},
+};
+export interface ProofOfMisbehaviorAsset {
+	readonly header1: BlockHeader;
+	readonly header2: BlockHeader;
+	// tslint:disable-next-line readonly-keyword
+	reward?: bigint;
+}
+
+export class ProofOfMisbehaviorTransaction extends BaseTransaction {
+	public readonly asset: ProofOfMisbehaviorAsset;
+	public static TYPE = 15;
+
+	public constructor(rawTransaction: unknown) {
+		super(rawTransaction);
+		const tx = (typeof rawTransaction === 'object' && rawTransaction !== null
+			? rawTransaction
+			: {}) as Partial<TransactionJSON>;
+		this.asset = (tx.asset || {}) as ProofOfMisbehaviorAsset;
+		this.asset.reward = BigInt(0);
+	}
+
+	public assetToJSON(): ProofOfMisbehaviorAsset {
+		return {
+			header1: this.asset.header1,
+			header2: this.asset.header2,
+		};
+	}
+
+	protected assetToBytes(): Buffer {
+		return Buffer.concat([
+			getBlockBytesWithSignature(this.asset.header1),
+			getBlockBytesWithSignature(this.asset.header1),
+		]);
+	}
+
+	public async prepare(store: StateStorePrepare): Promise<void> {
+		const delegateId = getAddressFromPublicKey(
+			this.asset.header1.generatorPublicKey,
+		);
+
+		const filterArray = [
+			{
+				address: this.senderId,
+			},
+			{
+				address: delegateId,
+			},
+		];
+
+		await store.account.cache(filterArray);
+	}
+
+	protected validateAsset(): ReadonlyArray<TransactionError> {
+		const asset = this.assetToJSON();
+		const schemaErrors = validator.validate(
+			proofOfMisbehaviorAssetFormatSchema,
+			asset,
+		);
+		const errors = convertToAssetError(
+			this.id,
+			schemaErrors,
+		) as TransactionError[];
+
+		if (
+			this.asset.header1.generatorPublicKey !==
+			this.asset.header2.generatorPublicKey
+		) {
+			errors.push(
+				new TransactionError(
+					'GeneratorPublickey of each blockheader must be matching.',
+					this.id,
+					'.asset.header1',
+				),
+			);
+		}
+
+		if (
+			Buffer.compare(
+				getBlockBytes(this.asset.header1),
+				getBlockBytes(this.asset.header2),
+			) === 0
+		) {
+			errors.push(
+				new TransactionError(
+					'Blockheaders are identical. No contradiction detected.',
+					this.id,
+					'.asset.header1',
+				),
+			);
+		}
+
+		/*
+            Check for BFT violations:
+                1. Double forging
+                2. Disjointness 
+                3. Branch is not the one with largest maxHeighPrevoted
```

Comment:
maybe it's better to separate the function of `checkHeadersContradicting(blockHeader1,blockHeader2)`, and write exactly as LIP is written?

Reply:
I feel it is unnecessary.